pytools/__init__.py

```diff
@@ -267,43 +271,83 @@ def delta(x, y):
         return 0
 
 
-def levi_civita(tup):
-    """Compute an entry of the Levi-Civita tensor for the indices *tuple*."""
+def levi_civita(tup: Tuple[int, ...]) -> int:
+    """Compute an entry of the Levi-Civita symbol for the indices *tuple*."""
     if len(tup) == 2:
         i, j = tup
-        return j-i
+        return j - i
     if len(tup) == 3:
         i, j, k = tup
-        return (j-i)*(k-i)*(k-j)/2
+        return (j-i) * (k-i) * (k-j) // 2
     else:
-        raise NotImplementedError
+        raise NotImplementedError(f"Levi-Civita symbol in {len(tup)} dimensions")
 
 
-def factorial(n):
-    from operator import mul
-    assert n == int(n)
-    return reduce(mul, (i for i in range(1, n+1)), 1)
+factorial = MovedFunctionDeprecationWrapper(math.factorial)
 
+try:
+    # NOTE: only available in python >= 3.8
+    perm = math.perm
+except AttributeError:
+    def _unchecked_perm(n, k):
+        result = 1
+        while k:
+            result *= n
+            n -= 1
+            k -= 1
 
-def perm(n, k):
-    """Return P(n, k), the number of permutations of length k drawn from n
-    choices.
-    """
-    result = 1
-    assert k > 0
-    while k:
-        result *= n
-        n -= 1
-        k -= 1
+        return result
 
-    return result
+    def perm(n: SupportsIndex, k: Optional[SupportsIndex] = None) -> int:
+        """
+        :returns: :math:`P(n, k)`, the number of permutations of length :math:`k`
+            drawn from :math:`n` choices.
+        """
+        from warnings import warn
+        warn("This function is deprecated and will go away in 2022. "
```

Comment:
Hmm. As it stands, the deprecation warning gets shown to folks with Python 3.7 and lower, which is approximately nobody. Want to add a `__getattr__`-based warning instead?

Reply:
@inducer Hm, yeah, I couldn't decide what's the best way to do this.
* Python >=3.8 gets a deprecation warning?
* Python <3.8 gets a deprecation warning + suggestion to update? There isn't anything to suggest to use instead here though.
* Both?

Reply:
https://github.com/inducer/pytools/compare/501b7a0f5178137e8fad1f4a0fc02f67d416d9ce..cd515dcb143974f2eb8f1fe132298bb3b7eea515
Thoughts?

Reply:
Ok, `mypy` seems pretty unhappy with that, but otherwise it looks good to me.